crates/net/downloaders/src/headers/linear.rs

```diff
@@ -1,427 +1,857 @@
     }
-}
 
-impl<C: Consensus, H: HeadersClient> LinearDownloader<C, H> {
-    fn new_download(&self, head: SealedHeader, tip: H256) -> HeadersDownload<C, H> {
-        HeadersDownload {
-            head,
-            tip,
-            buffered: VecDeque::default(),
-            request: Default::default(),
-            consensus: Arc::clone(&self.consensus),
-            request_retries: self.request_retries,
-            batch_size: self.batch_size,
-            client: Arc::clone(&self.client),
-            encountered_error: false,
+    /// Returns the next header request
+    ///
+    /// This will advance the current block towards the local head.
+    ///
+    /// Returns `None` if no more requests are required.
+    fn next_request(&mut self) -> Option<HeadersRequest> {
+        let local_head = self.local_block_number();
+        if self.next_request_block_number > local_head {
+            let request = calc_next_request(
+                local_head,
+                self.next_request_block_number,
+                self.request_batch_size,
+            );
+            // need to shift the tracked request block number based on the number of requested
+            // headers so follow-up requests will use that as start.
+            self.next_request_block_number -= request.limit;
+
+            return Some(request)
         }
+
+        None
     }
-}
 
-type HeadersFut = Pin<Box<dyn Future<Output = PeerRequestResult<BlockHeaders>> + Send>>;
+    /// Returns the next header to use for validation.
+    ///
+    /// Since this downloader downloads blocks with falling block number, this will return the
+    /// lowest (in terms of block number) validated header.
+    ///
+    /// This is either the last `queued_validated_headers`, or if has been drained entirely the
+    /// `lowest_validated_header`.
+    ///
+    /// This only returns `None` if we haven't fetched the initial chain tip yet.
+    fn lowest_validated_header(&self) -> Option<&SealedHeader> {
+        self.queued_validated_headers.last().or(self.lowest_validated_header.as_ref())
+    }
 
-/// A retryable future that returns a list of [`BlockHeaders`] on success.
-struct HeadersRequestFuture {
-    request: HeadersRequest,
-    fut: HeadersFut,
-    retries: usize,
-    max_retries: usize,
-}
+    /// Processes the next headers in line.
+    ///
+    /// This will validate all headers and insert them into the validated buffer.
+    ///
+    /// Returns an error if the given headers are invalid.
+    ///
+    /// Caution: this expects the `headers` to be sorted with _falling_ block numbers
+    #[allow(clippy::result_large_err)]
+    fn process_next_headers(
+        &mut self,
+        request: HeadersRequest,
+        headers: Vec<Header>,
+        peer_id: PeerId,
+    ) -> Result<(), HeadersResponseError> {
+        let mut validated = Vec::with_capacity(headers.len());
+
+        for parent in headers {
+            let parent = parent.seal();
```

Comment:
Would like to convert this to in parallel if possible? Maybe not using rayon, might require some multithreading from our side manually

Reply:
these are only 1000 items at a time, the start-up cost for parallelization should be higher.